### ISSUE TITLE
Exported enum - inspector selection breaks when enum is updated to contain additional values

### ISSUE THREAD

**Issue by Giwayume** [godotengine/godot#56000]
### Godot version

3.4.stable

### System information

Windows 7 64 bit

### Issue description

Creating enums with auto-assigned integer values ends up as a bad experience when exporting them for editing in the inspector, and you need to later update them. I could have 100s of instances of an object I made but I need to add a new enum value, and this ends up breaking all of my object instances because the actual assigned values of the enums changed.

If I'm using this syntax, that means I don't want to be concerned with the actual value of the enum, I only care that the enum name matches before the interpreter comes in to replace with an integer value.
```
enum SOME_ENUM {
    APPLE,
    PEAR,
    PEACH
}
```

But since the inspector seems to care about the integer value, not the enum name, you can end up with a mismatch of state if you add new values to the enum.

### Steps to reproduce

```
enum SOME_ENUM {
    APPLE,
    PEAR,
    PEACH
}

export(SOME_ENUM) var enum_selection
```

In the inspector, select PEACH for "Enum Selection".

Update SOME_ENUM to contain a value that comes before PEACH, such as:

```
enum SOME_ENUM {
    APPLE,
    GRAPE,
    PEAR,
    PEACH
}
```

Now "Enum Selection" will say "PEAR" instead of "PEACH".

I'm assuming this is because Godot stores the integer value for the enum, since the integer values are auto assigned in this case, the values increment.

**Comment by pycbouh** [comment]
You assume correctly, enums are not a type of data in GDScript, it’s just a list of constants (if unnamed), or a dictionary (if named).

**Comment by Giwayume** [comment]
I'd suggest the inspector store the enum name instead of value if possible. From a UI perspective, all it shows is the name of the enum. There's no hint or warning that this is a problem that could occur in this case.

![image](https://user-images.githubusercontent.com/4075314/146448840-5a31eee1-8b77-442b-afa9-ea578559f4c9.png)

**Comment by pycbouh** [comment]
Inspector doesn't "store" anything. It's merely a UI to display and allow you to edit exported properties of your classes. And enum is an integer, it is stored as an integer.

**Comment by Giwayume** [comment]
Sure, that doesn't change the fact that this is a really bad "gotcha" and can lead to a large amount of rework when encountered. This includes potential bugs due to the difficulty in identifying every single scene instance in every single scene.

Whether the mitigation is **improving** how the inspector keeps track of enums, or adding a gdscript warning when exporting enums without assigned values, my point in opening this is something should be done. An end user isn't going to inherently understand how the Godot editor works and why this back breaking situation might happen.

**Comment by pycbouh** [comment]
But this is unrelated to how the editor works... It will behave the same if you set the value in code, saved the scene, and then add a new enum value in between other values. That's just something you need to learn about working with enums, as in GDScript they are just numbers, nothing else.

Is it not obvious and can it lead to unfortunate mistakes? Sure. Does it make it harder to work with them because you need to account for shifting values? Yes, and you can only solve it by giving your enums enough space for future additions. But I'm not sure where we should warn about any of it, other than documentation. Not setting values to your enums is still a valid syntax and has valid uses, and you may not care about this particular circumstance.

**Comment by Giwayume** [comment]
> Not setting values to your enums is still a valid syntax and has valid uses, and you may not care about this particular circumstance.

Which is why I'd prefer if this can be fixed in the editor to store enum selection by name instead of value. Or store **both** in the .tscn files in case it changes it can still be tracked.

**Comment by Giwayume** [comment]
> But this is unrelated to how the editor works... 

This is 100% about the editor. In gdscript you always reference an auto-assigned enum by name, you don't care about the value. In the inspector, you select it in the dropdown by name. But the inspector will break in this scenario, while gdscript **will not**.

```
enum FRUITS {
    APPLE,
    PEAR
}

var my_fruit = FRUITS.PEAR

if my_fruit == FRUITS.PEAR:
    eat(my_fruit)
```

Add a new value to the FRUITS enum, gdscript will not break. This problem is all about the editor interaction with gdscript. Developer interaction through gdscript cares about the enum name, whereas the editor cares about the enum value.

**Comment by pycbouh** [comment]
> It will behave the same if you set the value in code, **save the scene**, and then add a new enum value in between other values.

Highlighted the important part that you've likely missed. Again, this is not related to the inspector. It is related to how enums are serialized, and them being integers means they are serialized as integers. If you store `my_fruit` in any way using Godot's built-in serialization methods, it will be stored as a number, and as such will be prone to the same issue if you change underlying enum and override the old value with a new name.

Inspector just displays it as it is stored.

**Comment by Giwayume** [comment]
You're really hyper focused on technicalities. I believe I've thoroughly pointed out the usability problem.

**Comment by pycbouh** [comment]
I've acknowledged usability problems, but they are inherent to the way enums work in Godot, and there is nothing specific about Inspector that is broken here. I still have no idea how and where we can warn users, because it is completely fine to use enums without specific numeric values, and not everyone will run into issues doing that. So reporting that as a potential problem by default is not going to make great user experience for a lot of users.

**Comment by pycbouh** [comment]
Completely inappropriate hypothetical aside... Okay, warning in the Inspector added. How do we warn people who store enums from code (say, as a part of a custom resource) and can run into the same issue?

**Comment by Giwayume** [comment]
The point of the hypothetical is to make you to ignore how it's currently implemented and focus on how it would be solved if it was a hard requirement. 

The other point your bring up is foreign to me, that seems like a separate issue. I personally store these types of enums by serializing their names into JSON.

**Comment by pycbouh** [comment]
Well, I'm trying to explain to you that it is not a separate issue. If you have custom serialization code, that's your decision, but that's not what we suggest users do in Godot, as we have native resources that fit the purpose much better. We can't solve half of a problem, and it's not great to spam people with warnings about things they may not care about, because they would add to the end of that enum, for example.

The problem here is only about learning how enums work, what enums are. Users working with code are expected to learn about those things, and consequences of writing particular code. Again, it is a gotcha, I've run into it a million times and worked around it all the same. I'm not rejecting it exists, but we cannot fix what you feel the issue is. We need to either fix what actually the issue is, or document the behavior better and ask people to learn the tech that they use.

**Comment by Giwayume** [comment]
Something in terms of a warning is going to be better than nothing, if you can't figure out an approach to fix the underlying issue, even if not all potential scenarios are covered. 

But this is where I differentiate, yes it is a programmer's job to learn the language. I disagree it's the game developer's job to learn how the Godot editor itself works behind the scenes to store values that you pick in the inspector. Those are two different things, and different expectations. You could have coded the editor in a non-generic way that uses a custom function for each value type instead of serialization, is it a game developer's job to know the ins and outs of that implementation?

**Comment by pycbouh** [comment]
> I like to sort my enums logically or alphabetically... most long term projects that will encounter this issue will have similar rules for organization in order to keep it maintainable.

And once you learn how enums work GDScript, you can add paddings and fix numerical values to specific names. It's a pretty normal practice with enums in any language, when the serialization or networking is involved.

> But this is where I differentiate, yes it is a programmer's job to learn the language. I disagree it's the game developer's job to learn how the Godot editor itself works behind the scenes to store values that you pick in the inspector. Those are two different things, and different expectations. You could have coded the editor in a non-generic way that uses a custom function to handle populating the dropdown for each value type instead of relying on serialization, is it a game developer's job to know the ins and outs of that implementation?

I'm going to sound like a broken record, but this is unrelated to the editor. The editor only shows what is stored, and what is stored is an integer. It is stored as an integer whether you edit it from the Inspector or from code. There is absolutely no difference, when it comes to serializing scenes or resources. In fact, enums can't even be used as types. If you write type-safe code, you would use integer variables.

Maybe you expect enums in GDScript to be enums from some other language, where they are a complete entity of their own. This is not the case in GDScript. Like I've said in the first reply, they are either an integer constant, if unnamed, or a dictionary of string keys and integer values, if named. They are just syntactic sugar over these, nothing more. They don't validate values, neither. You can store any integer in those fields, because that's what they are. Export hints are just that, hints to help display the property in the inspector. But it's still a numeric property.

If you want to store strings instead, you can do it with such hint for a String property:
```
export(String, "Apple,Pear,Banana") var fruit : String
```

You can also generate this list on the fly using `_get_property_list()`:
```
func _get_property_list() -> Array:
	var props := []
	
	props.append({
		"name": "my_fruit",
		"type": TYPE_STRING,
		"hint": PROPERTY_HINT_ENUM,
		"hint_string": "Apple,Pear,Banana"
	})
	
	return props
```

**Comment by Giwayume** [comment]
Then change the serialization? My point is, it doesn't matter if they're the same, you can't expect a user to know how the editor works behind the scenes.

**Comment by pycbouh** [comment]
We can't make an integer to be serialized as anything but an integer... If you want to have strings, you can use strings instead of enums.

**Comment by Giwayume** [comment]
> We can't make an integer to be serialized as anything but an integer

I highly doubt this, that's a personal decision. You can make a serialization to be whatever you want, it's a format you define. Even if it's "integer" vs "integer with extra metadata to include original enum name that you can ignore if you don't need".

**Comment by name-here** [comment]
@pycbouh It may by that fixing this would have unacceptable downsides, like ugly code or more maintenance work down the line, but it's certainly not impossible.  The question to answer isn't "why is this impossible?"  The question is "how would we do this, and would the usability improvement be worth the effort required to implement it?"

**Comment by pycbouh** [comment]
@name-here I was just explaining what the actual cause of the issue is, so that we can discuss a solution, a workaround for the user, or a documentation improvement. Unfortunately, I failed to reach the user as they were too focused on what the perceived issue is, weren't interested in a workaround, and didn't accept changes to documentation as a possible solution.

**Comment by Giwayume** [comment]
Um, this issue isn't about me. "Use strings or any assigned values instead of int as a workaround" - great, I know that. The issue is about the particular bug with auto assigned values and how to fix or warn about it. Those workarounds you mentioned are irrelevant to the case at hand.

And no one reads all the documentation before they start coding, that's not really a great solution. Many times people are following tutorial videos, not reading the official documentation.

I even see this recommendation everywhere on the forums, which would expose this bug:

```
enum DROPOFF { linear,square,none }
export(DROPOFF) var dropoff = DROPOFF.linear
```

Would probably be a good step to change the documentation to not use auto assigned enums everywhere they're used as an export, but you're still not going to catch everyone this way.

**Comment by pycbouh** [comment]
I agree, that people often tend to avoid documentation either due to their eagerness to start, or out of fear of not understanding theoretical part. And we can't quality control third-party content, like tutorials. But the docs should still cover such things, as their lacking is often the source of misconceptions and bad tutorials.

> make a huge project with it, then add new values to the enum, breaking all of the instances of the object in all of the scenes they've place it in, causing a mess. "You could've done it this way" isn't much help when such disastrous damage has already been done.

That's not really an issue specific to this gotcha with enums actually being integers. If you make a huge project using tech you don't quite know that well, you will run into such issues. We can't cancel the learning curve, and one shouldn't overestimate their abilities. Especially in gamedev.

I hope you don't still see my objections as being dismissive. That's a part of my job to try and find a solution that is both technically feasible and actually improves user experience. And the issue you raise here is old and familiar, but we are not yet able to find a good solution, at least not until we can have enums to be actual enums, and not just a syntactic sugar on top of integer constants and dictionaries.

And I don't find the idea of a warning every time you have such enum exported to be a particularly good idea, because it would have false negatives just as well as current lack of any warning has false positives. And it doesn't really help solve the problem if the user never exports the enum, but still uses it in serialization.

Serializing it as a string-int pair also has its problems, as you wouldn't be able to validate the value without loading and parsing the corresponding script. So you see, that gets to a non-trivial, core touching territory very fast. Which is why an improvement to the docs may just be a solution that we need now.

**Comment by Giwayume** [comment]
> If you make a huge project using tech you don't quite know that well, you will run into such issues.

Yeah, that's the one part I disagree with, I don't think it's a gamedev's job to understand absolutely everything about a technology before they start using it, it should be reasonable building blocks. Especially since the editor itself should be a "black box" that "just works". No other game engine out there expects you to learn how the editor stores your selection behind the scenes in order to prevent issues like this, or even asks you to learn how your assets are serialized. In many cases like Unreal Engine they **don't want you to know**, it's a secret.

**Comment by Giwayume** [comment]
Are enums not changed in 4.0 to be proper enums? I never looked at that.

**Comment by pycbouh** [comment]
> Yeah, that's the one part I disagree with, I don't think it's a gamedev's job to understand absolutely everything about a technology before they start using it, it should be reasonable building blocks. 

I'm not saying it's your job as a developer to know everything about the tool, but the less you are experienced, the more gotchas are going to happen. It's just a fact of using any tool, and you either have to embrace creating and shipping code that works even if it's ugly or not well maintainable, or you end up with a dead project.

> Especially since the editor itself should be a "black box" that "just works".

I will tirelessly repeat that this has nothing to do with the editor, but you will tireless ignore it, right? I mean, at this point, how else can I explain to you that this is not an editor issue, never was an editor issue, and cannot be solved just as an editor issue? It's an issue with the underlying typing system of the engine, and GDScript in particular, that doesn't have an enum entity, and isn't aware of any enums in a practical sense.

> No other game engine out there expects you to learn how the editor stores your selection behind the scenes in order to prevent issues like this, or even asks you to learn how your assets are serialized. In many cases like Unreal Engine they don't want you to know, it's a secret.

That's just not true. You often need to know how that works, especially if you are making the aforementioned "huge projects".

> Are enums not changed in 4.0 to be proper enums? I never looked at that.

There were some changes, but they are still not there yet. They can be used for typing, but it's not enforced and any numeric value would be accepted by the property.

> Is parsing the corresponding script not already being done on the editor side in order to populate the dropdown list?

By the time the editor gets involved, everything should already be parsed. Because this is not related to the editor. Deserialization of scenes is a part of the engine. It's how your projects work in general. When the scene is loaded by the engine, it would need to load and run the script to validate that enum value, but it also needs to do it before assigning it to the property, because there is no enum property, there is only int property. Then there is a question of whether the name of the stored value should be trusted, or the value.

Only long after the scene was loaded, validated and restored does the editor appear in this picture. If it appears at all, because the exact same logic would have to happen whenever you load a scene regardless in the game, or in a plugin.

**Comment by Giwayume** [comment]
> You often need to know how that works, especially if you are making the aforementioned "huge projects".

I mean, in my example with Unreal Engine they don't make this information available. It is open source but secret through obscurity, likely to not encourage game hacks and stolen assets. Pretty much only game modders have reverse engineered the serialization format through years of work, they're just finally getting around this year to being able to parse serialized Unreal 4 assets outside of the engine's code. So no, the game developer would not have a chance to learn this, they have to trust that the engine's mechanisms just work.

> When the scene is loaded by the engine, it would need to load and run the script to validate that enum value, but it also needs to do it before assigning it to the property, because there is no enum property, there is only int property. Then there is a question of whether the name of the stored value should be trusted, or the value.

If going down this road I'd say the name is used if the current variable value doesn't correspond to an existing enum key that has an explicitly assigned value. Otherwise the name is ignored. The problem isn't that value is used, it's when value is auto-assigned and user doesn't exactly care what the value is, so name would be used instead in that case.

And based on your description, it doesn't sound like storing the extra data for the enum name is really that much additional CPU work at all since the enum is already being parsed by the interpreter. You'd only need to store and worry about matching up based on name if it was an auto assigned enum value selected in the first place.
